handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java

```diff
@@ -151,14 +151,24 @@
         SET_SERVER_NAMES_METHOD = setServerNamesMethod;
     }
 
-    private static final int MAX_PLAINTEXT_LENGTH = 16 * 1024; // 2^14
-    private static final int MAX_COMPRESSED_LENGTH = MAX_PLAINTEXT_LENGTH + 1024;
-    private static final int MAX_CIPHERTEXT_LENGTH = MAX_COMPRESSED_LENGTH + 1024;
+    static final int MAX_PLAINTEXT_LENGTH = 16 * 1024; // 2^14
 
-    // Header (5) + Data (2^14) + Compression (1024) + Encryption (1024) + MAC (20) + Padding (256)
-    static final int MAX_ENCRYPTED_PACKET_LENGTH = MAX_CIPHERTEXT_LENGTH + 5 + 20 + 256;
+    /**
+     * This is the maximum overhead when encrypting plaintext as defined by
+     * <a href="https://www.ietf.org/rfc/rfc5246.txt">rfc5264</a>,
+     * <a href="https://www.ietf.org/rfc/rfc5289.txt">rfc5289</a> and openssl implementation itself.
+     *
+     * Please note that we use a padding of 16 here as openssl uses PKC#5 which uses 16 bytes while the spec itself
+     * allow up to 255 bytes. 16 bytes is the max for PKC#5 (which handles it the same way as PKC#7) as we use a block
+     * size of 16. See <a href="https://tools.ietf.org/html/rfc5652#section-6.3">rfc5652#section-6.3</a>.
+     *
+     * 16 (IV) + 48 (MAC) + 1 (Padding_length field) + 16 (Padding) + 1 (ContentType) + 2 (ProtocolVersion) + 2 (Length)
```

Comment:
There's also AEAD-based ciphers which have a completely different derivation, but they all have much smaller overhead than this. I don't believe you need the +1. There is a length, but since the goal is to get the total size to a multiple of the block size, minimal padding will only take you up at most one block anyway.